File: Top-Interview-Questions-Leetcode/Easy-Collections/Array/plus-one.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<int> plusOne(std::vector<int>& digits) {
        int n = digits.size();
        int r = 0;
        
        digits[n - 1] += 1;
        
        for (int i = n - 1; i > 0; --i) {
            digits[i] += r;
            
            if (digits[i] > 9) {
                digits[i] %= 10;
                r = 1;
            }
            else r = 0;
        }
        
        digits[0] += r;
        
        if (digits[0] > 9) {
            digits[0] %= 10;
            digits.insert(digits.begin(), 1, 1);
        }
        
        return digits;
    }
};

// Example usage
int main() {
    Solution solution;
    std::vector<int> digits = {9, 9, 9, 9, 9, 9};
    
    for (int digit : digits) {
        std::cout << digit;
    }

    std::cout << " + 1 = ";

    std::vector<int> result = solution.plusOne(digits);

    for (int digit : result) {
        std::cout << digit;
    }
    
    return 0;
}

File: Top-Interview-Questions-Leetcode/Easy-Collections/Array/plus-one.py
from typing import List

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        n, r = len(digits), 0
        
        digits[-1] += 1
        
        for i in range(n - 1, 0, -1):
            digits[i] += r
            r = 0
            
            if digits[i] > 9:
                digits[i] %= 10
                r = 1
        
        digits[0] += r
        
        if digits[0] > 9:
            digits[0] %= 10
            digits.insert(0, 1)
        
        return digits

# Example usage
def main():
    digits = [9, 9, 9]
    print(f"{''.join(map(str, digits))} + 1 = ", end="")
    solution = Solution()
    result = solution.plusOne(digits)
    print(f"{''.join(map(str, result))}")

if __name__ == "__main__":
    main()